packages/babel-plugin-transform-inline-consecutive-adds/src/index.js

```diff
@@ -0,0 +1,161 @@
+  // otherwise returns [foo as string, ?rval, index of the variable declaration]
+
+  const declarations = path.get("declarations");
+  if (declarations.length !== 1) {
+    return;
+  }
+
+  const declaration = declarations[0];
+  const id = declaration.get("id"), init = declaration.get("init");
+  if (!id.isIdentifier()) {
+    return;
+  }
+
+  const parent = path.parentPath;
+  if (!parent.isBlockParent() || !parent.isScopable()) {
+    return;
+  }
+
+  const body = parent.get("body");
+  const startIndex = body.indexOf(path);
+  if (startIndex === -1) {
+    return;
+  }
+
+  return [id.node.name, init, startIndex];
+}
+
+function collectExpressions(path, checkExprType) {
+  // input: ExprStatement => "a | SequenceExpression
+  // SequenceExpression => "a list
+  // Validates "a is of the right type
+  // returns null if found inconsistency, else returns Array<"a>
+  if (path.isExpressionStatement()) {
+    const exprs = collectExpressions(path.get("expression"), checkExprType);
+    return (exprs !== null) ? exprs : null;
+  }
+
+  if (path.isSequenceExpression()) {
+    const exprs = path.get("expressions")
+                      .map((p) => collectExpressions(p, checkExprType));
+    if (exprs.some((e) => e === null)) {
+      return null;
+    } else {
+      return exprs.reduce((s, n) => s.concat(n), []);  // === Array.flatten
+    }
+  }
+
+  if (checkExprType(path)) {
+    return [path];
+  }
+
+  return null;
+}
+
+function getContiguousStatementsAndExpressions(body, start, end, checkExprType, checkExpr) {
+  const statements = [];
+  let allExprs = [];
+  for (let i = start; i < end; i++) {
+    const exprs = collectExpressions(body[i], checkExprType);
+    if (exprs === null || !exprs.every((e) => checkExpr(e))) {
+      break;
+    }
+    statements.push(body[i]);
+    allExprs = allExprs.concat(exprs);
+  }
+  return [statements, allExprs];
+}
+
+
+module.exports = function({ types: t }) {
+  return {
+    name: "transform-inline-consecutive-adds",
+    visitor: {
+      VariableDeclaration(path) {
+        const topLevel = validateTopLevel(path);
+        if (topLevel === null) {
+          return;
+        }
+
+        const [name, init, startIndex] = topLevel;
+        const references = getIdAndFunctionReferences(name, path.parentPath);
+        const body = path.parentPath.get("body");
+
+        for (let collapser of collapsers) {
+          if (!collapser.checkInitType(init)) {
+            continue;
+          }
+
+          const [statements, exprs] = getContiguousStatementsAndExpressions(
+            body,
+            startIndex + 1,
+            body.length,
+            collapser.checkExpressionType,
+            collapser.makeCheckExpression(name, references)
+          );
+
+          if (statements.length > 0) {
+            const addons = exprs.map((e) => collapser.extractAddon(e));
+            addons.forEach((addon) => collapser.addAddon(t, addon, init));
+            statements.forEach((s) => s.remove());
```

Comment:
I really like the overall architecture here, where we have a nice modular way of defining extractable behavior for various data types — arrays, objects, sets, whatever else comes in the future.

But I did find some names a little confusing while reading it and trying to figure it out.

I would rename addons to assignments — `extractAssignment`, `addAssignment` — since addons is a little ambiguous and feels out of place.

I also found `makeCheckExpression` a little confusing. How about `makeExpressionValidator` (alternatively `getExpressionValidator` or `make/getExpressionChecker`) and `checkExpressionType` --> `isExpressionTypeValid`